File: backend/schemas/product.py
# backend/schemas/product.py
"""
Product schemas for API request/response validation.

FIX (2026-02-21): Added PlatformLink and platforms_linked to ProductRead
to support multi-platform badges in the product list UI. See BUG-005.
"""

from datetime import datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

# ============== Request Schemas ==============


class ProductCreate(BaseModel):
    name: str = Field(..., max_length=255)
    sku: str | None = Field(default=None, max_length=100)
    description: str | None = None
    base_price: Decimal = Field(..., gt=0)
    category: str | None = Field(default=None, max_length=100)
    image_url: str | None = None
    is_active: bool = True
    cost: Decimal | None = Field(default=None, ge=0, description="Cost to acquire/produce")
    min_price: Decimal | None = Field(default=None, gt=0)
    max_price: Decimal | None = Field(default=None, gt=0)
    sentiment_multiplier: Decimal = Field(default=Decimal("0.1"), ge=0, le=1)
    auto_pricing_enabled: bool = False
    keywords: list[str] = []


class ProductUpdate(BaseModel):
    name: str | None = Field(default=None, max_length=255)
    category: str | None = Field(default=None, max_length=100)
    image_url: str | None = None
    is_active: bool | None = None
    sku: str | None = Field(default=None, max_length=100)
    description: str | None = None
    base_price: Decimal | None = Field(default=None, gt=0)
    current_price: Decimal | None = Field(default=None, gt=0)
    cost: Decimal | None = Field(default=None, ge=0)
    min_price: Decimal | None = Field(default=None, gt=0)
    max_price: Decimal | None = Field(default=None, gt=0)
    sentiment_multiplier: Decimal | None = Field(default=None, ge=0, le=1)
    auto_pricing_enabled: bool | None = None
    keywords: list[str] | None = None


# ============== Response Schemas ==============


class PlatformLink(BaseModel):
    """Platform connection info for a product."""

    platform: str  # "shopify" or "woocommerce"
    store_url: str | None = None
    external_price: float | None = None
    sync_enabled: bool = False


class ProductRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    user_id: UUID
    name: str
    sku: str | None
    description: str | None
    category: str | None
    image_url: str | None
    is_active: bool
    base_price: Decimal
    current_price: Decimal
    cost: Decimal | None
    min_price: Decimal | None
    max_price: Decimal | None
    sentiment_multiplier: Decimal
    auto_pricing_enabled: bool
    keywords: list[str]
    created_at: datetime
    updated_at: datetime
    # FIX BUG-005: Platform connection info
    platforms_linked: list[PlatformLink] = []


# ============== Price Suggestion Schema ==============


class PriceSuggestion(BaseModel):
    product_id: UUID
    current_price: Decimal
    suggested_price: Decimal
    change_percent: Decimal
    reasoning: str
    confidence: Decimal = Field(ge=0, le=1)
    factors: dict

File: frontend/types/product.ts
// Product domain types
// AUTO-SYNCED with backend via openapi-typescript
// Last synced: 2026-01-08
// Source: components["schemas"]["ProductCreate"], ProductRead, ProductUpdate

// ============================================
// PRODUCT TYPES - Matches backend exactly
// ============================================

/**
 * Product response from GET endpoints
 * Matches: components["schemas"]["ProductRead"]
 */
export interface Product {
  id: string;
  user_id: string;
  name: string;
  sku: string | null;
  description: string | null;
  category: string | null;
  image_url: string | null;
  is_active: boolean;
  base_price: string;        // Decimal returned as string
  current_price: string;     // Decimal returned as string
  cost: string | null;       // Decimal returned as string
  min_price: string | null;  // Decimal returned as string
  max_price: string | null;  // Decimal returned as string
  sentiment_multiplier: string; // Decimal returned as string
  auto_pricing_enabled: boolean;
  keywords: string[];
  created_at: string;        // ISO 8601 datetime
  updated_at: string;        // ISO 8601 datetime
}

/**
 * Create product request
 * Matches: components["schemas"]["ProductCreate"]
 * 
 * Note: Decimal fields accept number | string for flexibility
 * Backend Pydantic will coerce to Decimal
 */
export interface CreateProductRequest {
  name: string;
  sku?: string | null;
  description?: string | null;
  base_price: number | string;  // Required - accepts both
  category?: string | null;
  image_url?: string | null;
  is_active?: boolean;          // Default: true
  cost?: number | string | null;
  min_price?: number | string | null;
  max_price?: number | string | null;
  sentiment_multiplier?: number | string; // Default: 0.1
  auto_pricing_enabled?: boolean;         // Default: false
  keywords?: string[];                    // Default: []
}

/**
 * Update product request
 * Matches: components["schemas"]["ProductUpdate"]
 * All fields optional for PATCH semantics
 */
export interface UpdateProductRequest {
  name?: string | null;
  sku?: string | null;
  description?: string | null;
  base_price?: number | string | null;
  current_price?: number | string | null;
  category?: string | null;
  image_url?: string | null;
  is_active?: boolean | null;
  cost?: number | string | null;
  min_price?: number | string | null;
  max_price?: number | string | null;
  sentiment_multiplier?: number | string | null;
  auto_pricing_enabled?: boolean | null;
  keywords?: string[] | null;
}

// ============================================
// PAGINATED RESPONSE
// ============================================

/**
 * Paginated products response
 * Matches: components["schemas"]["PaginatedResponse_ProductRead_"]
 */
export interface PaginatedProducts {
  items: Product[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
  total_pages?: number;  // Alias for pages
}

// ============================================
// PRICE SUGGESTION (AI-generated)
// ============================================

/**
 * Price suggestion from AI endpoint
 * Note: This is returned from /products/{id}/suggestion
 */
export interface PriceSuggestion {
  product_id: string;
  current_price: number;
  suggested_price: number;
  change_percent: number;
  reasoning: string;
  confidence: number;
  factors: {
    sentiment_score: number | null;
    competitor_avg_price?: number | null;
    mention_count?: number;
    mention_volume?: number;
    trend?: string;
    ai_key_factors?: string[];
    ai_powered?: boolean;
  };
}

// ============================================
// PRICE HISTORY
// ============================================

/**
 * Price history entry
 * Matches: components["schemas"]["PriceHistoryResponse"]
 */
export interface PriceHistoryEntry {
  id: string;
  product_id: string;
  price: string;
  source: string;
  change_reason: string | null;
  created_at: string;
}

// ============================================
// PRODUCT SUMMARY (Dashboard)
// ============================================

/**
 * Product summary for dashboard cards
 * Matches: components["schemas"]["ProductSummary"]
 */
export interface ProductSummary {
  id: string;
  name: string;
  sku?: string | null;
  current_price: string;
  base_price: string;
  price_change_percent: number;
  sentiment_score?: number | null;
  mention_count_24h: number;        // Default: 0
  has_pending_recommendation: boolean; // Default: false
  auto_pricing_enabled: boolean;    // Default: false
}
